placement/algo/algo.go

```diff
@@ -75,45 +76,61 @@ func (a rackAwarePlacementAlgorithm) RemoveHost(ps placement.Snapshot, leavingHo
 }
 
 func (a rackAwarePlacementAlgorithm) AddHost(ps placement.Snapshot, addingHost placement.Host) (placement.Snapshot, error) {
-	var addingHostShards placement.HostShards
-	var err error
-	if addingHostShards, err = getNewHostShardsForSnapshot(ps, addingHost); err != nil {
-		return nil, err
-	}
-	return a.addHostShards(ps, addingHostShards)
+	ps = placement.CopySnapshot(ps)
+	return a.addHostShards(ps, placement.NewHostShards(addingHost))
 }
 
-func (a rackAwarePlacementAlgorithm) ReplaceHost(ps placement.Snapshot, leavingHost, addingHost placement.Host) (placement.Snapshot, error) {
-	ph, leavingHostShards, addingHostShards, err := newReplaceHostPlacementHelper(a.options, ps, leavingHost, addingHost)
+func (a rackAwarePlacementAlgorithm) ReplaceHost(ps placement.Snapshot, leavingHost placement.Host, addingHosts []placement.Host) (placement.Snapshot, error) {
+	ps = placement.CopySnapshot(ps)
+	ph, leavingHostShards, addingHostShards, err := newReplaceHostPlacementHelper(ps, leavingHost, addingHosts, a.options)
 	if err != nil {
 		return nil, err
 	}
 
-	var shardsUnassigned []uint32
 	// move shards from leaving host to adding host
-	for _, shard := range leavingHostShards.Shards() {
-		if moved := ph.MoveShard(shard, leavingHostShards, addingHostShards); !moved {
-			shardsUnassigned = append(shardsUnassigned, shard)
+	hh := ph.BuildHostHeap(addingHostShards, true)
+	for leavingHostShards.ShardsLen() > 0 {
+		if hh.Len() == 0 {
+			break
+		}
+		tryHost := heap.Pop(hh).(placement.HostShards)
+		if moved := ph.MoveOneShard(leavingHostShards, tryHost); moved {
+			heap.Push(hh, tryHost)
 		}
 	}
 
-	// if there are shards that can not be moved to adding host
-	// distribute them to the cluster
-	if err := ph.PlaceShards(shardsUnassigned, leavingHostShards); err != nil {
-		return nil, err
+	if leavingHostShards.ShardsLen() > 0 && !a.options.AllowPartialReplace() {
+		return nil, fmt.Errorf("could not fully replace all shards from %s, %v shards left unassigned", leavingHost.ID(), leavingHostShards.ShardsLen())
 	}
 
-	// add the adding host to the cluster and bring its load up to target load
-	cl := ph.GenerateSnapshot()
+	// place the shards from the leaving host to the rest of the cluster
+	if err := ph.PlaceShards(leavingHostShards.Shards(), leavingHostShards); err != nil {
+		return nil, err
+	}
+	// fill up to the target load for added hosts if have not already done so
+	newPS := ph.GenerateSnapshot()
+	for _, addingHost := range addingHosts {
+		newPS, leavingHostShards, err = removeHostFromPlacement(newPS, addingHost)
+		if err != nil {
+			return nil, err
+		}
+		newPS, err = a.addHostShards(newPS, leavingHostShards)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
yeah so addHostShards() would add a HostShards object to the placement and bring it's capacity to target load, this loop here is for cases when some of the new hosts could not reach it's ideal target load due to, say, rack conflict, when taking shards from the leaving host. This operation here would move some shards from some other loaded hosts to this under loaded host. 

The removeHostFromPlacement() will not remove all the shards from the HostShards object, it will just remove it from the placement so I can call addHostShards() to add it back again.

At this moment, we will only be setting a.options.AllowPartialReplace() = false, so we will never do this shard distribution optimization. But in the future this could help maintaining a evenly distributed cluster